src/cli_util.rs

```diff
@@ -1229,6 +1231,13 @@ pub struct GlobalArgs {
         default_value = "@"
     )]
     pub at_operation: String,
+    /// Enable verbose logging
+    #[arg(long, short = 'v', global = true, help_heading = "Global Options")]
+    pub verbose: bool,
+}
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct EarlyArgs {
```

Comment:
Why do we need to split out `EarlyArgs`? Will any of the other arguments in `GlobalArgs` cause problems if we allow them early?

Reply:
> Will any of the other arguments in GlobalArgs cause problems if we allow them early?

Yes, which is why the `EarlyArgs` are split out. Some examples from my early testing are..

- `no_commit_working_copy`. This one is easy, we can replace bool with `Option<bool>` like we do for `no_pager`.
- `at_operation`. This is a `String` that defaults to `"@"`, not so easy to replace.

An alternative is to have an `Args::new()` that initializes the struct to the default values and use `update_from_args`, but this requires us to duplicate the default values in the clap definition and `Args::new()` (Unless there's some elegant way to reuse the default values off clap, which I doubt; https://docs.rs/clap/latest/clap/builder/struct.Arg.html#method.get_default_values returns `&[OsStr]`, which makes it somewhat unusable for `bool`).

Reply:
I was thinking that we could parse the arguments we care about while ignoring the rest. Something similar to what we do when resolving aliases here: https://github.com/martinvonz/jj/blob/47067c1368a3d3e5831744f00ee5d1d18b70b271/src/cli_util.rs#L1352. Hopefully there's a way of allowing unknown flags too. Would that work?

Reply:
> I was thinking that we could parse the arguments we care about while ignoring the rest

Even in that code snippet you shared, `try_get_matches_from` will throw if we encounter parsing errors.
AFAICT the only options for partial parsing are:

-`ignore_errors` and don't use `from_arg_matches`
- parse using a struct with only the arguments we care about

(https://github.com/clap-rs/clap/issues/1880 is the thread for partial parsing)
In either case, we'd still need `ignore_errors` to handle help, because clap treats help like an error.

Reply:
I see. What I didn't understand before was that you want color to be disabled if the user does `jj --color=never --nonexistent`. I suppose it would still work with the current definition of `GlobalArgs` because those arguments cannot currently be invalid, but if we added two mutually exclusive flags to `GlobalArgs`, then we would no longer be able to use it, so it's better to extract the `EarlyArgs` already.

Reply:
> you want color to be disabled if the user does jj --color=never --nonexistent

Ah, yes, but the normal case is even more troublesome. e.g. if you have `jj --color=never help`, clap will see the `help` and immediately short-circuit in `try_get_matches_from()`. You can't work around this with `disable_help_subcommand()`, because then clap will short-circuit on the missing subcommand.

To bypass both of these problems, you need to pass `ignore_errors()`, which _should_ be the end of it, but because of the no default values bug, we need to make silly adjustments to the args that we want to be parsed early, and because of those adjustments, we don't want to parse all args early.

> if we added two mutually exclusive flags to GlobalArgs, then we would no longer be able to use it

I'm not _too_ worried about this, since it's reasonable to error out if the user asks for something conflicting.

Reply:
Yes, though you'd also need to disable explicit help via subcommand and arg, so the matching would be something like:

```
    let early_matches = app.clone()
        .disable_help_subcommand(true)
        .disable_help_flag(true)
        .allow_external_subcommands(true)
        .tbd_allow_unkown_flags_or_something(true) // I haven't found a setting that does this yet.
        .get_matches();
```

I think the only blocker is allowing the now-unknown `--help` flag (I don't know if the option exists), but it should allow us to leave `GlobalArgs` untouched (in theory at least).

Reply:
Do we have the same problem with the `--help` flag when using `EarlyArgs`? Either way, I don't worry too much if it seems hard to use `GlobalArgs` untouched. I think it makes sense to extract `EarlyArgs` anyway

Reply:
> Do we have the same problem with the --help flag when using EarlyArgs?

I assume you meant `ignore_errors()`, which is the thing that works around `--help` and `help` (not `EarlyArgs`). We don't have that problem, and presumably `ignore_errors()` would work around all other things too (like `--version`, if that ever matters).